### PR TITLE
Validate grant types

### PR DESCRIPTION
**Pull request by tfrommen**
Hi again,

@rmccue didn't comment on my _What's Left_ list in #8, but since this was no big deal, I just created this pull request with a proposed validation for grant type( handler)s.

The result of the grant type filter, `oauth2.grant_types`, will be validated, and only `WP\OAuth2\Types\Type` implementations make it to the consumer.

Comments?

Cheers,
Thorsten

### CODE REVIEW

**Comment by rmccue** [comment]
Whoops, let me go back and answer :)

File: plugin.php [hunk 1]
```diff
@@ -73,9 +74,14 @@ function get_grant_types() {
...
 	 * OAuth 2 specification.
 	 *
-	 * @param array $grant_types Map of grant type to handler object.
+	 * @param Type[] $grant_types Map of grant type to handler object.
 	 */
-	return apply_filters( 'oauth2.grant_types', array() );
+	$grant_types = apply_filters( 'oauth2.grant_types', array() );
+
+	return array_filter( $grant_types, function ( $type ) {
+
```
Comment (rmccue) [inline plugin.php]: No need for the blank line here.

Comment (tfrommen) [inline plugin.php]: Removed (aka _fixed_).

**Comment by tfrommen** [comment]
>Could we instead just specify function ( Type $type ) and have PHP validate for us? 🤔

No, not really. PHP would not **validate** (and discard invalid) types, like the code that I suggested, but rather **expect** correct implementations, and fatal error if encountered anything else.
If this is the desired behavior, then sure, we could add a type hint. But the `array_filter()` would then get pretty awkward to look at, in my opinion. We do not want to do anything to valid types, so the code could look something like this:

```php
return array_filter( $types, function ( Type $type )  {
    return true;
} );
```

So, in fact, we wouldn't really need the callback('s body), but just PHP that kicks in for checking (and fatal erroring because of) the type.

To be honest, I would rather keep my code (or something along the lines). But like I said, the bahavior would be _discard invalid types_. If you want PHP to jump the user into their face, we can do that. 😀 

Does this make sense? 🤔

**Comment by rmccue** [comment]
I'd rather not silently drop objects; it makes debugging a pain for developers. If not the fatal error, we should at least fire a `_doing_it_wrong` when we drop it.

**Comment by tfrommen** [comment]
Makes sense. Should I adapt the closure accordingly?

**Comment by rmccue** [comment]
That'd be great, thanks :)

**Comment by tfrommen** [comment]
I just updated the validation code, again.

If only we could use PHP 5.6+... :)
I decided to use a regular `foreach` loop now, because then I'm able to include the type in the message. If this is not required, I can refactor this to use `array_filter()`. Another option would be to use `array_walk()` in combination with another variable, or `array_filter()` - which both might not be faster than the `foreach`.

Comments?

**Pull request merged**